[AppConfig] Fix disposed stream access

jsquire opened the pull request:
# Summary

The focus of these changes is to fix a bug where the content stream returned by the protocol method has been disposed and the serializer attempts to access it.  Shifting to accessing the `BinaryData` content directly for deserialization.

# References and Resources

- [[BUG] ObjectDisposedException when calling GetConfigurationSettingAsync / SetConfigurationSettingAsync (#38219)](https://github.com/Azure/azure-sdk-for-net/issues/38219)

annelo-msft committed:
DynamicData: Allow arrays of objects where elements are allowed types (#38458)

* Allow arrays of objects that are allowed types

* cast to string array instead

jsquire committed:
[AppConfig] Fix disposed stream access

The focus of these changes is to fix a bug where the content stream returned
by the protocol method has been disposed and the serializer attempts to access
it.  Shifting to accessing the `BinaryData` content directly for deserialization.

jsquire commented on sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationMockTests.cs:
```diff
@@ -15,6 +17,7 @@
 using Azure.Core;
 using Azure.Core.Pipeline;
 using Azure.Core.TestFramework;
+using Microsoft.VisualBasic;
```
```suggestion
```

jsquire committed:
Update sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationMockTests.cs

uh, WTF?

annelo-msft commented on sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs:
```diff
@@ -21,13 +19,16 @@ public partial class ConfigurationClient
         private static async Task<Response<ConfigurationSetting>> CreateResponseAsync(Response response, CancellationToken cancellation)
         {
-            ConfigurationSetting result = await ConfigurationServiceSerializer.DeserializeSettingAsync(response.ContentStream, cancellation).ConfigureAwait(false);
+            using var contentStream = response.Content.ToStream();
```
One observation is that this creates a new MemoryStream under the hood.  Depending on how concerned we are with performance in this scenario, we could update the DeserializeSetting method to take a BinaryData.  If it's just using it to parse a JsonDocument, it should be a small change, and if you change the method signature, you might catch other instances of this bug while you're in there.

jsquire replied:
I've got some time to play with this one.   I'll poke around a bit and see how much churn moving to `BinaryData` would cause.

annelo-msft commented on sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@
 
 - Marked a constructor overload of `ConfigurationSetting` that was intended for testing purposes as non-visible, as the `ConfigurationModelFactory` should instead be used.
 
+- Fixed a bug where a disposed content stream was used to attempt deserialization in some scenarios, such as using a custom `HttpMessageHandler`.
```
Curious: do we know why we only hit it when we set a customer message handler and not in other cases?  It feels like that would be good to understand.

jsquire replied:
I believe this is because of the Core  HttpClientTransport and its PipelineResponse], which manipulates the content and replaces the content stream.   I haven't been able to trace it all the way through, though.

jsquire replied:
This is related to the type of content set on the `HttpResponseMessage`.  

When `HttpMessage` is disposed, it disposes the `HttpResponseMessage` which, in turn, disposes its content.  In the case of `StringContent` and others derived from `HttpContent` the reference to the content stream is assumed to be owned by the content and disposed with it.

When forming the `PipelineResponse`, Core will read content stream and if it is a MemoryStream then it is considered buffered and kept without making a copy.  If it is another type of stream - such as a network stream - it is read into a MemoryStream.   

Because Core did not copy the stream read from `StringContent` set by the custom handler, it was disposed with `HttpMessage` despite appearing in `PipelineResponse` as a separate member.

annelo-msft replied:
Oh killer.  Great investigation, @jsquire!  Would it make sense to update the buffering policy to handle this HttpContent type?  I know we've duplicated some of the buffering logic outside that policy, e.g. [here](https://github.com/Azure/azure-sdk-for-net/pull/35716/files#diff-9916010ee7060f5a4ec225b148221bfa381b0f235a4f9f4a740900f44439aaffR230) so we might need to fix it in multiple places and/or find a way to consolidate the "should I buffer" logic to a single place.  None of that needs to block this PR, though, but it would be good to file an issue to track that work if you haven't already.  Thank you!

pallavit replied:
In theory this issue can happen in any SDK that has a similar pattern so shouldn't the right fix\onus of buffering\creating the copy be on the core pipeline pieces?

jsquire replied:
The current generation doesn't have this problem because the response is processed in a scope where the `HttpMessage` is still alive.  Still not 100% sure whether this is a special case because it's old hand-written code being integrated into the DPG surface or if there's a legitimate pattern fix needed here.

I'd agree that we should consider revisiting the `PipelineResponse` approach and consider some alternate way to signal "this is buffered content owned by the pipeline" rather than just making assumptions based on the stream type.

That said, I don't have a deep familiarity with the pipeline implementation and would like to discuss with the Core team before we decide next steps.    If I'm correct, I believe this is a potential issue for all protocol methods.

Regardless, the AppConfig implementation is a hybrid approach due to the history.  We should still adjust the consumption pattern here.

pallavit replied:
Perfect! My uber concern is if this is a legit DPG pattern we should move the fix up in Core than having each SDK fend for itself.

annelo-msft replied:
> My uber concern is if this is a legit DPG pattern

I do think this is an issue of crossing the HLC/DPG streams, as Jesse indicated.  It would be good to have an issue filed in Core to track investigating whether we have a deeper problem here, if this is a use case we want to support.  I'm happy to let @jsquire drive it as he sees fit, though.

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

stephentoub committed:
Fix broken chat functions example in README (#38461)

* Fix broken chat functions example in README

The use of the ChatRole.Function message will fail if the message doesn't have its Name set to the name of function.

* Update test as well

annelo-msft committed:
Tidy up some Core shared source files (#38447)

* Remove LLPH and update PH

* remove another unused file

* restore PageableHelpers

benbp committed:
Use identity subscription config and pre script for live tests (#38473)

* Use dedicated identity subscription config for live tests

* Move identity variable setting into test-resources-pre.ps1

* Handle local testing mode

christothes committed:
Automate Managed Identity manual tests - Azure Functions (#38393)

trrwilson committed:
readme: add functions/extensions compat note (#38475)

azure-sdk committed:
Update Generator Version 3.0.0-beta.20230901.2 (#38478)

ethanann-ms committed:
Users/ethanann/swagger package preview 2023 06 (#38247)

azure-sdk committed:
Increment package version after release of Azure.ResourceManager.IotHub (#38484)

ArcturusZhang committed:
prepare release for mobile network (#38488)

azure-sdk committed:
Update AutoRest C# version to 3.0.0-beta.20230904.1 (#38487)

Mohana-Krishna-Nali committed:
Managed Network Fabric - Updates ] (#38054)

* Supporting new default optional params in  ACL and RoutePolicy

* Minor change

* Removing the trailing space

* Resolving the errors in the PR

* Minor change

* Merging the branch with the latest main.

* Mitigate breaking changes

* Export API

* Added scenarios for
ACLS
NetworkPacketBroker
NetworkTapRule
NetworkTap
NeighbourGroup

* Updated the resource details in scenario based on test execution

* Fixed the autorest.md and regenerating the code.

* Code cleanup

* Auo generated api changes

* Update code added in Internet Gateway tests

* Updated assert json

* Updating the non real subscription

* Updated asserts json

* Updated subscription values

* Addressing review comments

* Updated the version in change log md file

* Addressed review commetns

* Updated version in Managed Network project file

* prepare release

---------

Co-authored-by: Wei Hu (from Dev Box) <huwe@microsoft.com>
Co-authored-by: Saravanakumar Mylsamy <smylsamy@microsoft.com>

azure-sdk committed:
Update Generator Version 3.0.0-beta.20230905.1 (#38494)

azure-sdk committed:
Increment package version after release of Azure.ResourceManager.MobileNetwork (#38493)

KalpeshChavan12 committed:
Added InGuestPatch Test case. (#38483)

christothes committed:
VM managed identity integration test (#38457)

azure-sdk committed:
Increment package version after release of Azure.ResourceManager.Maintenance (#38499)

jsquire committed:
changed synchronous deserialization to use memory rather than a stream to avoid the wrapper.  Asynchronus deserialziation requires a stream, so the wrapper is, unfortuantely, still needed.

jsquire commented:
Well, that rebase did not go well.   Closing this out to avoid the mass inclusion.

jsquire commented:
Superseded by #38531, which fixes the bad rebase.
